ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -996,12 +1017,74 @@ let rank_hosts_by_best_vgpu ~__context vgpu visible_hosts =
                0L
         )
         hosts
-      |> List.map (fun g -> List.map (fun (h, _) -> h) g)
+      |> List.map (fun g -> List.map fst g)
+
+(* Group all hosts to 2 parts:
+   1. Home server. A list of home server (only one host).
+   2. A list of lists, each list contains hosts with the same count of running VM in the same anti-affinity group.
+      These lists are sorted by VM's count.
+   Combine these lists into one list. The list is like below:
+   [[host1] (home server), [host2, host3] (no VM running), [host4, host4] (one VM running), [host5, host6] (more VMs running), ...]
+*)
+let rank_hosts_by_placement ~__context ~vm ~group =
+  let home_server_hosts =
+    match Db.VM.get_affinity ~__context ~self:vm with
+    | ref when Db.is_valid_ref __context ref ->
+        [ref]
+    | _ ->
+        []
+  in
+  let resident_hosts =
+    Db.VM_group.get_VMs ~__context ~self:group
+    |> List.filter_map (fun vm ->
+           let host =
+             Db.VM.get_scheduled_to_be_resident_on ~__context ~self:vm
+           in
+           if host <> Ref.null then
+             Some host
+           else
+             let host = Db.VM.get_resident_on ~__context ~self:vm in
+             if
+               host <> Ref.null
+               && Db.VM.get_power_state ~__context ~self:vm = `Running
```

Comment:
When a VM.resident_on is set, I understand any non-running power state would be a transient state. So maybe this checking could be ignored.

Reply:
When a VM is paused, I see its `resident_on` is not cleared:
```
[root@genuk-21-05d ~]# xe vm-param-list uuid=33436b42-a21c-e25a-4d78-ddca795da344 | grep resident-on
                           resident-on ( RO): b46a9faa-da91-4bde-b7be-ae4db174d1a9
[root@genuk-21-05d ~]#
[root@genuk-21-05d ~]# xe vm-pause vm=rhel8_43
[root@genuk-21-05d ~]# xe vm-param-list uuid=33436b42-a21c-e25a-4d78-ddca795da344 | grep resident-on
                           resident-on ( RO): b46a9faa-da91-4bde-b7be-ae4db174d1a9
[root@genuk-21-05d ~]#
[root@genuk-21-05d ~]# xe vm-unpause vm=rhel8_43
[root@genuk-21-05d ~]#
[root@genuk-21-05d ~]# xe vm-shutdown vm=rhel8_43

[root@genuk-21-05d ~]#
[root@genuk-21-05d ~]# xe vm-param-list uuid=33436b42-a21c-e25a-4d78-ddca795da344 | grep resident-on
                           resident-on ( RO): <not in database>

```

Reply:
I mean could a VM paused on a host be counted in the sense of VM anti-affinity?

Reply:
In requirement, the paused VM should not be considered when computing anti-affinity:
https://info.citrite.net/display/xenserver/Requirement+for+XenServer+VM+anti-affinity+rules+support#:~:text=without%20breaking%20them.-,SP3,-.%20The%20anti%2Daffinity